runtime/src/accounts_db.rs

```diff
@@ -2329,26 +2324,28 @@ impl<'a> AppendVecScan for ScanState<'a> {
         self.pubkey_to_bin_index -= self.bin_range.start;
 
         let balance = loaded_account.lamports();
-        let loaded_hash = loaded_account.loaded_hash();
-        let source_item = CalculateHashIntermediate::new(loaded_hash, balance, *pubkey);
+        let mut loaded_hash = loaded_account.loaded_hash();
 
-        if self.config.check_hash
+        let hash_is_missing = loaded_hash == Hash::default();
+        if (self.config.check_hash || hash_is_missing)
             && !AccountsDb::is_filler_account_helper(pubkey, self.filler_account_suffix)
         {
-            // this will not be supported anymore
             let computed_hash = loaded_account.compute_hash(
                 self.current_slot,
                 pubkey,
-                INCLUDE_SLOT_IN_HASH_IRRELEVANT_CHECK_HASH,
+                self.config.include_slot_in_hash,
             );
-            if computed_hash != source_item.hash {
+            if hash_is_missing {
+                loaded_hash = computed_hash;
```

Comment:
Maybe assert `computed_hash != Hash::default()` or later assert `loaded_hash != Hash::default()` to ensure missing hash case is always handled?

Reply:
in theory, Hash::default() is a valid hash value, though obviously unlikely. If we fail to hash a missing hash, we'll find out by subsequent failures.

Reply:
> in theory, Hash::default() is a valid hash value, though obviously unlikely.

Yep, that's true 😅.  If the hash is already the default, then recomputing it should always result in default as well.